### PR TITLE
226 add parity to fertility module

### PR DESCRIPTION
**Pull request by paddy-r**
~Tested with ```make fertility_testing``` and ```make S7_intervention_hhIncome```, and already merged with ```development``` to resolve conflicts.~

Testing with:
```
make baseline_default
make S7_baseline
make cv_default
make cv_SIPHER7
```

...plus all R notebooks producing reasonable results.

### CODE REVIEW

File: minos/data_generation/US_missing_LOCF.py
Comment (ld-archer) [inline minos/data_generation/US_missing_LOCF.py:1]: Just need to change the docstring for this, was written with education in mind but being used for nkids_ind now as well.

File: minos/minosPipeline/RunPipeline.py [hunk 1]
```diff
@@ -48,6 +48,100 @@
...
+replenishment_components_map = {
+    "Replenishment()": Replenishment(),
+    "NoReplenishment()": NoReplenishment(),
+    "ReplenishmentNowcast()": ReplenishmentNowcast(),
+    "ReplenishmentScotland()": ReplenishmentScotland(),
+}
+
+
+# HR 19/04/23 Rough definition of component priorities
+# Replenishment component first, then intervention(s), then fert/mort, then everything else
```
Comment (ld-archer) [inline minos/minosPipeline/RunPipeline.py:102]: Think the order should be slightly different here. Should be:
1. Replenishment
2. Fert/Mort
3. Income
4. Intervention
5. Everything else

File: minos/modules/replenishment.py [hunk 2]
```diff
@@ -66,6 +66,9 @@ def setup(self, builder):
                         'loneliness',
                         'weight',
                         'nkids',
+                        'nkids_ind',
+                        'nkids_ind_raw',
+                        'nkids_ind_new',
```
Comment (ld-archer) [inline minos/modules/replenishment.py:71]: Are `nkids_ind_raw` or `nkids_ind_new` used in the fertility module or rest of the pipeline? If not (or if any aren't used) they might be better being removed from the dataframe before entering the pipeline.

File: minos/utils.py [hunk 3]
```diff
@@ -151,4 +196,638 @@ def patch(module, modification_function=replace_text, *args, **kwargs):
...
     sys.modules[module_name] = module
     globals()[module_name] = module
-    return module
\ No newline at end of file
+    return module
+
+
+# HR 26/05/23 To dump some population data for fertility/parity development testing
+# Grabs population at runtime, currently in block in add_new_birth_cohorts.nkidsFertilityAgeSpecificRates.on_time_step
+def dump_parity(pop,
```
Comment (ld-archer) [inline minos/utils.py:204]: Is this function still needed? Can it be generalised into something that would be useful for later on? If no to both maybe remove.

File: scripts/Makefile [hunk 4]
```diff
@@ -65,6 +65,9 @@ intervention_livingWage: setup
 intervention_energyDownLift: setup
 	$(PYTHON) scripts/run.py -c $(RUN_CONFIG) -o $(MODE) -i 'energyDownlift'
 
+fertility_testing: setup
```
Comment (ld-archer) [inline scripts/Makefile:68]: Target still needed? Remove if not.

**Review by ld-archer — CHANGES_REQUESTED** [review]
All testing good and changes look fine. I've added some comments to certain scripts mostly for stuff that I don't think is needed anymore, with one other change to the ordering of modules that needs a little tweak. Should be good to merge when you've addressed the comments, cheers.

File: minos/data_generation/US_missing_LOCF.py
Comment (paddy-r) [reply minos/data_generation/US_missing_LOCF.py:1]: Done in [e5fb2d2](https://github.com/Leeds-MRG/Minos/pull/259/commits/e5fb2d2acb8a8dbbe678d4fdb496a78fe8efcceb)

File: scripts/Makefile (hunk 4 above)
Comment (paddy-r) [reply scripts/Makefile]: Done in [a90b58e](https://github.com/Leeds-MRG/Minos/pull/259/commits/a90b58e06cb51816b7a34545c3919d61da54337a)

File: minos/utils.py (hunk 3 above)
Comment (paddy-r) [reply minos/utils.py]: Done in [35e4b3e](https://github.com/Leeds-MRG/Minos/pull/259/commits/35e4b3ed14843de42405612838c71485801f7dec)

File: minos/minosPipeline/RunPipeline.py (hunk 1 above)
Comment (paddy-r) [reply minos/minosPipeline/RunPipeline.py:102]: Need to discuss best way to do this, pending merge and resolving conflicts.

**Review by RobertClay — APPROVED** [review]
Working for me. See no glaring issues in outputs or anything on top of Lukes change requests.

File: minos/minosPipeline/RunPipeline.py (hunk 1 above)
Comment (paddy-r) [reply minos/minosPipeline/RunPipeline.py]: Can't do until after merge but some prep in [ad45938](https://github.com/Leeds-MRG/Minos/pull/259/commits/ad45938ff3eb87c2fd389fc5d86309bb73fd9f48)

Comment (paddy-r) [reply minos/minosPipeline/RunPipeline.py]: Fixed in [be59252](https://github.com/Leeds-MRG/Minos/pull/259/commits/be59252fd0b968d39d8b0284826a3f1bb4da8529)

**Pull request merged**